### ISSUE TITLE
Split up the script crate for build speed

### ISSUE THREAD

**Issue by kmcallister** [servo/servo#1799]
It's by far the largest crate in terms of binary size (2.5× as big as `main`).  Building it takes almost 40 seconds and 1.8 GB resident on my machine (with debugging enabled, optimization disabled).

The various DOM types and their generated bindings mostly don't depend on each other.  We could split them among a few crates to improve compile time (especially in parallel).

**Comment by jdm** [comment]
Splitting the script crate in a linear fashion may be more feasible. If we were able to move all of the generated bindings into a crate that script depends upon, that would mean that bindings and layout could build in parallel, and would reduce the working memory required to compile the rest of script. It would also speed up making changes to non-codegen parts of script.

**Comment by antrik** [comment]
I'd like to point out that this issue is more serious nowadays, as not only does the `script` crate totally dominate partial rebuild time (it takes several minutes for a non-optimised build now), but also the memory usage has gotten so large that it actually makes debug builds exceed the `3 GiB` virtual memory available on 32 bit systems, as reported in https://github.com/servo/servo/issues/13045 -- and it's probably just a matter of time until non-debug builds break as well...

**Comment by Ms2ger** [comment]
I don't see how that would work. The generated bindings need to be able to inline the method definitions on the concrete type, and the concrete type depends fundamentally on the wrap function, and the class / JSNative definitions with it.

**Comment by jdm** [comment]
Latest idea from discussions with @asajeffrey:
* split generated bindings code into two crates - code that can be parameterized over FooTraits types (to avoid needing to import concrete DOM types), and code that needs to refer to the parameterized code using concrete types
* extract low-level types from script/ that the generated bindings code needs to refer to (e.g. converting native types to JS values) and put them in the generic bindings crate

This should give us the following crates:
js_bindings (generic codegen, generated types (eg. unions, dictioanries), common script/ types)
script (DOM API implementations and non-generic codegen)

This would mean that non-WebIDL/CodegenRust.py changes should not have to rebuild a significant portion of the generated bindings code.

**Comment by jdm** [comment]
Things that will make the above idea difficult:
* concrete DOM types in generated types (eg. unions or dictionaries containing WebIDL interface types)

**Comment by jdm** [comment]
At minimum, this will require collecting the list of WebIDL interfaces transitively used by a given WebIDL property (for each return value and argument), adding generic parameters for each, and propagating these generic types through layers like the union conversion methods when interacting with them.

**Comment by jdm** [comment]
Potential solution to the previous comment: a single trait with associated types for every WebIDL interface, and every generic method gets parameterized over the trait and uses the T::Interface rather than Interface.

**Comment by asajeffrey** [comment]
Or even better (?) make each FooMethods trait parametric, rather than each method inside FooMethods. Something like:

trait DOMTypes {
  type Foo: FooMethods<Self>;
  ...
}
trait FooMethods<DOM: DOMTypes> {
  fn bar(x: DOM::Bar) -> DOM::Baz;
  ...
}
struct DOM;
impl DOMTypes for DOM {
  type Foo = Foo;
  ...
}

struct Foo { ... }
impl FooMethods<DOM> for Foo { ... }

**Comment by jdm** [comment]
FooMethods would be unchanged. It's the standalone binding glue functions that need to be parametric.

**Comment by asajeffrey** [comment]
How can FooMethods remain unchanged? Is the plan that FooMethods would be generated in the same crate as the concrete implementations? That does mean that the DOMTypes trait couldn't mention the FooMethods trait, but I don't know if that matters.

**Comment by jdm** [comment]
Oh right, FooMethods has arguments and return types too. Hooray!

**Comment by Ms2ger** [comment]
I wonder if we can use `-> impl BarTypes` now.

**Comment by asajeffrey** [comment]
We have quite a lot of places in the script crate where we assume the WebIDL method returns the concrete type, e.g. that `document.Window()` returns a `Root<Window>` not an `impl WindowMethods`.

**Comment by jdm** [comment]
My recommendation for how to go about this - to sort out fundamental problems quickly, I propose copying the codegen infrastructure into a new project, taking some representative WebIDL files and DOM implementations, and doing whatever it necessary to make it build unmodified. At that point the rebuild time should be _way_ smaller than Servo, which will yield much quicker feedback on experiments.

**Comment by asajeffrey** [comment]
IRC conversation with @nox and @Wafflespeanut: http://logs.glob.uno/?c=mozilla%23servo&s=8+Mar+2017&e=8+Mar+2017#c626952\

TL;DR:

a) splitting script is speculative, it might or might not work,
b) it might all be rendered moot by rustc improving incremental computation,
c) it would be a truck load of work,
and d) we love having this debate.

**Comment by antrik** [comment]
@asajeffrey b) is not entirely true: while incremental compilation *might* help with the major pain point of long build times after small changes, initial build time is also somewhat relevant; and even more importantly, it won't help with the memory usage problem.

(Maybe other compiler improvements could help with the latter, though?...)

**Comment by asajeffrey** [comment]
True, I was mainly thinking about speeding up recompiling when editing DOM .rs and .webidl files. You can tell which case hurts me!

**Comment by nox** [comment]
I don't think this is useful to keep this issue open. We know that splitting the script crate is something we can consider, but this will be done so far in the future (if ever) that I am hoping by then incremental compilation will make the crate size a non-issue.

**Comment by antrik** [comment]
@nox I want to point out, again, that incremental compilation will *not* make the memory footprint during build a non-issue.

**Comment by asajeffrey** [comment]
The problem is still a real one, even if we don't have a clear vision of how to solve it, so I'd like to keep this issue open. If incremental computation makes this a non-issue, we can close the issue then.

**Comment by SimonSapin** [comment]
>  Building it takes almost 40 seconds

Hahaha, the good old times

**Comment by metux** [comment]
Any news on this issue ? Just having crashes by >4G memory consumption.

**Comment by jdm** [comment]
No news.

**Comment by metux** [comment]
hmm, I'll try with kicking out stuff I dont wanna have anyways (bluetooth, gamepad, vr, ...).
unfortunately, #[cfg(...)] doesn't seem to work in many places - I really miss a real preprocessor.

**Comment by haxxorsid** [comment]
Is the effort needed small, medium or xlarge?

**Comment by jdm** [comment]
The effort is still very large. I have proposed a [Google Summer of Code project](https://github.com/servo/servo/wiki/Prototype-ways-of-splitting-the-script-crate) to investigate the ideas we have discussed in this thread.

**Comment by sagudev** [comment]
While trying to understand the idea, work that was already done, its problems and how associate types would be used for this I make this example:

```rust
mod script_bindings {
    // this module can be considered as forward declaratins
    // but in servo it should also contain all generated stuff from codegen (uses)
    // while actual defieinitions/implementations is outside

    pub trait TypeHolderTrait: Sized {
        type Dom: DomTrait<Self>;
        type WGPU: WgpuTrait<Self>;
    }

    // all exposed functions need to be behind trait
    // this is not a problem for webidl exposed methods
    // as the trait for them is generated by codegen
    pub trait DomTrait<TH: TypeHolderTrait> {
        fn method_on_dom(&self) -> &TH::WGPU;
    }

    pub trait WgpuTrait<TH: TypeHolderTrait> {
        fn method_on_wgpu(&self) -> &TH::Dom;
    }
}

mod script {
    use crate::script_bindings::TypeHolderTrait;
    use crate::script_bindings::DomTrait;

    pub struct ActualDom<TH: TypeHolderTrait> {
        wgpu: TH::WGPU,
    }

    /// This method is not exposed via traits,
    /// so users needs to depend on script directly
    ///
    /// See `uses_no_trait` for usage
    impl<TH: TypeHolderTrait> ActualDom<TH> {
        pub fn no_trait(&self) {

        }
    }

    impl<TH: TypeHolderTrait> DomTrait<TH> for ActualDom<TH> {
        fn method_on_dom(&self) -> &TH::WGPU {
            &self.wgpu
        }
    }

    pub fn start<TH: TypeHolderTrait>() {
        // ...
    }
}

mod script_webgpu {
    use crate::script_bindings::TypeHolderTrait;
    use crate::script_bindings::WgpuTrait;

    pub struct ActualWGPU<TH: TypeHolderTrait> {
        dom: TH::Dom,
    }

    /// Here we use associated type binding syntax to tell compiler
    /// to use script::ActualDom for Dom type
    /// this allows us to call methods on ActualDom that were not "forward declered" via traits
    impl<TH: TypeHolderTrait<Dom = crate::script::ActualDom<TH>>> ActualWGPU<TH> {
        pub fn uses_no_trait(&self) {
            self.dom.no_trait()
        }
    }

    impl<TH: TypeHolderTrait> WgpuTrait<TH> for ActualWGPU<TH> {
        fn method_on_wgpu(&self) -> &TH::Dom {
            &self.dom
        }
    }
}

/// This links implementations from script-* with "forward declarations" from script_bindings
pub struct TypeHolder;
impl script_bindings::TypeHolderTrait for TypeHolder {
    type Dom = script::ActualDom<Self>;
    type WGPU = script_webgpu::ActualWGPU<Self>;
}

fn main() {
    // when starting script we just need to put puzzle together
    script::start::<TypeHolder>();
}
```

and I must say I am thrilled. I think it should be feasible to separate codegen bindings/actual implementations and later we could even split actual implementations (see `script_webgpu`)

**Comment by jdm** [comment]
I'm going to try hacking together a really minimal script_bindings that uses real codegen on one interface, just to try to tease out what changes are necessary.

**Comment by sagudev** [comment]
Well we already have https://github.com/servo/servo/pull/21371, I was thinking to go big and separate all generated code with implementations (`TypeHolderTrait` would generated by codegen). The main question I have is how to land such change (due to size).

**Comment by jdm** [comment]
https://github.com/servo/servo/compare/main...jdm:servo:script-split is the branch I worked on while flying on planes today. It successfully compiles with `RUSTC_BOOTSTRAP=1`; why does https://github.com/servo/servo/blob/main/components/script/lib.rs#L5 work for the script crate but fails to compile in the new script_bindings crate?

This feels promising! The codegen in script_bindings is spitting out this monster trait for my test webidl collection:
```rust
pub trait DomTypes: crate::utils::DomHelpers<Self> + Sized {
    type CSSStyleDeclaration: crate::dom::bindings::codegen::Bindings::CSSStyleDeclarationBinding::CSSStyleDeclarationMethods<Self> + crate::reflector::DomObject + crate::reflector::DomGlobal<Self> + crate::conversions::IDLInterface + PartialEq + crate::reflector::DomObjectWrap<Self> + crate::inheritance::Castable + malloc_size_of::MallocSizeOf + crate::reflector::MutDomObject + crate::conversions::DerivedFrom<Self::CSSStyleDeclaration>;
    type DOMException: crate::dom::bindings::codegen::Bindings::DOMExceptionBinding::DOMExceptionMethods<Self> + crate::reflector::DomObject + crate::reflector::DomGlobal<Self> + crate::conversions::IDLInterface + PartialEq + crate::reflector::DomObjectWrap<Self> + crate::inheritance::Castable + malloc_size_of::MallocSizeOf + crate::reflector::MutDomObject + crate::conversions::DerivedFrom<Self::DOMException>;
    type DOMParser: crate::dom::bindings::codegen::Bindings::DOMParserBinding::DOMParserMethods<Self> + crate::reflector::DomObject + crate::reflector::DomGlobal<Self> + crate::conversions::IDLInterface + PartialEq + crate::reflector::DomObjectWrap<Self> + crate::inheritance::Castable + malloc_size_of::MallocSizeOf + crate::reflector::MutDomObject + crate::conversions::DerivedFrom<Self::DOMParser>;
    type Document: crate::dom::bindings::codegen::Bindings::DocumentBinding::DocumentMethods<Self> + crate::reflector::DomObject + crate::reflector::DomGlobal<Self> + crate::conversions::IDLInterface + PartialEq + crate::reflector::DomObjectWrap<Self> + crate::inheritance::Castable + malloc_size_of::MallocSizeOf + crate::reflector::MutDomObject + crate::conversions::DerivedFrom<Self::Document>;
    type GlobalScope: crate::dom::bindings::codegen::Bindings::GlobalScopeBinding::GlobalScopeMethods<Self> + crate::reflector::DomObject + crate::reflector::DomGlobal<Self> + crate::conversions::IDLInterface + PartialEq + crate::reflector::DomObjectWrap<Self> + crate::inheritance::Castable + malloc_size_of::MallocSizeOf + crate::reflector::MutDomObject + crate::conversions::DerivedFrom<Self::GlobalScope>;
    type Window: crate::dom::bindings::codegen::Bindings::WindowBinding::WindowMethods<Self> + crate::reflector::DomObject + crate::reflector::DomGlobal<Self> + crate::conversions::IDLInterface + PartialEq + crate::reflector::DomObjectWrap<Self> + crate::inheritance::Castable + malloc_size_of::MallocSizeOf + crate::reflector::MutDomObject + crate::conversions::DerivedFrom<Self::GlobalScope> + crate::conversions::DerivedFrom<Self::Window>;
}
```
Then individual traits get parameterized like:
```rust
pub trait DOMParserMethods<D: DomTypes> {
    fn ParseFromString(&self, str: DOMString, type_: SupportedType) -> Fallible<DomRoot<D::Document>>;
    fn Constructor(global: &D::Window, proto: Option<HandleObject>) -> Result<DomRoot<D::DOMParser>, Error>;
}
```
And any bindings code that needs to interact with non-IDL methods for concrete DOM types needs to have the implementation provided by the script crate:
```
pub trait DomHelpers<D: crate::codegen::DomTypes::DomTypes> {
    fn throw_dom_exception(
        cx: crate::script_runtime::JSContext,
        global: &D::GlobalScope,
        result: crate::error::Error,
    );

    fn global_scope_from_object(obj: *mut JSObject) -> crate::root::DomRoot<D::GlobalScope>;

    fn global_scope_origin(global: &D::GlobalScope) -> &servo_url::MutableOrigin;
}
```

**Comment by sagudev** [comment]
> It successfully compiles with RUSTC_BOOTSTRAP=1; why does https://github.com/servo/servo/blob/main/components/script/lib.rs#L5 work for the script crate but fails to compile in the new script_bindings crate?

Because we set `RUSTC_BOOTSTRAP` for specific crates here:
https://github.com/servo/servo/blob/2db9032e72bda6108b4eb5988eaf164141a61d64/.cargo/config.toml#L31-L33

**Comment by sagudev** [comment]
> And any bindings code that needs to interact with non-IDL methods for concrete DOM types needs to have the implementation provided by the script crate

or use associated type binding syntax as shown above (will be needed for every consumer, so this is not appropriate for very used methods).

**Comment by sagudev** [comment]
> > And any bindings code that needs to interact with non-IDL methods for concrete DOM types needs to have the implementation provided by the script crate
> 
> or use associated type binding syntax as shown above (will be needed for every consumer, so this is not appropriate for very used methods).

For non-IDL methods that are impl on concrete DOM types we can use specific traits for for concrete DOM types (ex: `GlobalScopeHelpers` that would be implemented on `GlobalScope`), to add this to traits that are impl on `DomTypes::GlobalScope`, we can use new Bindings.conf field:
```conf
'GlobalScope': {
    'traits': ['GlobalScopeHelpers',],
},
```
and user will have to manually define it in core script crate:
```rust
pub trait GlobalScopeHelpers<D: crate::codegen::DomTypes::DomTypes<GlobalScope = Self>> {
    fn global_scope_from_object(obj: *mut JSObject) -> DomRoot<Self>;

    fn global_scope_origin(global: &Self) -> &servo_url::MutableOrigin;
}
```

functions that are not methods should simply be defined in core script with generic arg `throw_dom_exception<D: DomTypes>(...)`.

**Comment by sagudev** [comment]
I have created a document with all stuff in one place (and some more ideas) on how to do this: https://gist.github.com/sagudev/c4a029cfccda9553adac32749a9ba583

**Comment by jdm** [comment]
I am working on getting the script_bindings crate in my branch compiling with all real webidl files present. It's going surprisingly well so far.

**Comment by sagudev** [comment]
> I am working on getting the script_bindings crate in my branch compiling with all real webidl files present. It's going surprisingly well so far.

I was actually planing to do this, after I make sure that we can also measure runtime performance, hence my work on speedometer.

**Comment by jdm** [comment]
My script_bindings crate is now compiling with all DOM interfaces present \o/ Next step is getting the script crate to use tho generated bindings. https://github.com/jdm/servo/tree/script-split has all the latest changes.

**Comment by jdm** [comment]
I've made good progress in that branch, and I've found ways to make the changes in script/ minimally invasive, but I'm now running into errors that make no sense and feel like some kind of compiler bug. There are thousands of errors where the compiler can't resolve normal impl method calls, and it's hard to figure out how to narrow down what's going on.

**Comment by jdm** [comment]
Like, seriously. What is going on here?
```
error[E0599]: no method named `chars` found for reference `&str` in the current scope
   --> components/script/textinput.rs:247:19
    |
247 |     for c in text.chars() {
    |                   ^^^^^
    |
    = help: items from traits can only be used if the trait is in scope
help: trait `TextSource` which provides `chars` is implemented but not in scope; perhaps you want to import it
    |
7   + use unicode_bidi::TextSource;
    |
help: there is a method `par_chars` with a similar name
    |
247 |     for c in text.par_chars() {
    |                   ~~~~~~~~~
```

**Comment by sagudev** [comment]
let me try

**Comment by sagudev** [comment]
I got:
```
servo/target/debug/build/script-607bfb53410c9870/out/InterfaceTypes.rs: No such file or directory (os error 2)
   --> components/script/dom/mod.rs:209:5
    |
209 |     include!(concat!(env!("OUT_DIR"), "/InterfaceTypes.rs"));
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: this error originates in the macro `include` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `script` (lib) due to 1 previous error
```
on https://github.com/jdm/servo/commit/b05a845b3881f287907be101d9363b7d3500f37f. So you might need to run `mach clean` to make sure rust is not using old generated files.

**Comment by jdm** [comment]
Same errors after addressing that.

**Comment by sagudev** [comment]
Usually it's good idea to start looking at first errors and fixing them first (as those can cause more false errors later).

For me these are error due to orphaning rules:
```
error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
   --> components/script/dom/bluetoothuuid.rs:627:1
    |
627 | impl Clone for StringOrUnsignedLong {
    | ^^^^^^^^^^^^^^^--------------------
    | |              |
    | |              `script_bindings::codegen::UnionTypes::StringOrUnsignedLong` is not defined in the current crate
    | impl doesn't use only types from inside the current crate
    |
    = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
  --> components/script/dom/gpucanvascontext.rs:42:1
   |
42 | impl Clone for GPUCanvasConfiguration {
   | ^^^^^^^^^^^^^^^----------------------
   | |              |
   | |              `script_bindings::codegen::Bindings::WebGPUBinding::GPUCanvasConfiguration` is not defined in the current crate
   | impl doesn't use only types from inside the current crate
   |
   = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
  --> components/script/dom/gpucanvascontext.rs:54:1
   |
54 | impl Clone for HTMLCanvasElementOrOffscreenCanvas {
   | ^^^^^^^^^^^^^^^----------------------------------
   | |              |
   | |              `script_bindings::codegen::UnionTypes::HTMLCanvasElementOrOffscreenCanvas` is not defined in the current crate
   | impl doesn't use only types from inside the current crate
   |
   = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
   --> components/script/dom/request.rs:876:1
    |
876 | impl Clone for HeadersInit {
    | ^^^^^^^^^^^^^^^-----------
    | |              |
    | |              `script_bindings::codegen::UnionTypes::ByteStringSequenceSequenceOrByteStringByteStringRecord` is not defined in the current crate
    | impl doesn't use only types from inside the current crate
    |
    = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
  --> components/script/dom/gpucanvascontext.rs:63:1
   |
63 | impl malloc_size_of::MallocSizeOf for GPUTextureDescriptor {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^--------------------
   | |                                     |
   | |                                     `script_bindings::codegen::Bindings::WebGPUBinding::GPUTextureDescriptor` is not defined in the current crate
   | impl doesn't use only types from inside the current crate
   |
   = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
  --> components/script/dom/gpucanvascontext.rs:86:1
   |
86 | impl malloc_size_of::MallocSizeOf for HTMLCanvasElementOrOffscreenCanvas {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^----------------------------------
   | |                                     |
   | |                                     `script_bindings::codegen::UnionTypes::HTMLCanvasElementOrOffscreenCanvas` is not defined in the current crate
   | impl doesn't use only types from inside the current crate
   |
   = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
   --> components/script/security_manager.rs:170:1
    |
170 | impl Serialize for SecurityPolicyViolationEventDisposition {
    | ^^^^^^^^^^^^^^^^^^^---------------------------------------
    | |                  |
    | |                  `SecurityPolicyViolationEventDisposition` is not defined in the current crate
    | impl doesn't use only types from inside the current crate
    |
    = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
   --> components/script/dom/audiobuffersourcenode.rs:264:1
    |
264 | impl<'a> From<&'a AudioBufferSourceOptions> for AudioBufferSourceNodeOptions {
    | ^^^^^^^^^----------------------------------^^^^^----------------------------
    | |        |                                      |
    | |        |                                      `AudioBufferSourceNodeOptions` is not defined in the current crate
    | |        `script_bindings::codegen::Bindings::AudioBufferSourceNodeBinding::AudioBufferSourceOptions` is not defined in the current crate
    | impl doesn't use only types from inside the current crate
    |
    = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
   --> components/script/dom/audionode.rs:370:1
    |
370 | impl From<ChannelCountMode> for ServoMediaChannelCountMode {
    | ^^^^^----------------------^^^^^--------------------------
    | |    |                          |
    | |    |                          `servo_media::servo_media_audio::node::ChannelCountMode` is not defined in the current crate
    | |    `script_bindings::codegen::Bindings::AudioNodeBinding::ChannelCountMode` is not defined in the current crate
    | impl doesn't use only types from inside the current crate
    |
    = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
   --> components/script/dom/audionode.rs:380:1
    |
380 | impl From<ChannelInterpretation> for ServoMediaChannelInterpretation {
    | ^^^^^---------------------------^^^^^-------------------------------
    | |    |                               |
    | |    |                               `servo_media::servo_media_audio::node::ChannelInterpretation` is not defined in the current crate
    | |    `script_bindings::codegen::Bindings::AudioNodeBinding::ChannelInterpretation` is not defined in the current crate
    | impl doesn't use only types from inside the current crate
    |
    = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
   --> components/script/dom/audioparam.rs:325:1
    |
325 | impl From<AutomationRate> for ParamRate {
    | ^^^^^--------------------^^^^^---------
    | |    |                        |
    | |    |                        `ParamRate` is not defined in the current crate
    | |    `AutomationRate` is not defined in the current crate
    | impl doesn't use only types from inside the current crate
    |
    = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
   --> components/script/dom/baseaudiocontext.rs:591:1
    |
591 | impl From<ProcessingState> for AudioContextState {
    | ^^^^^---------------------^^^^^-----------------
    | |    |                         |
    | |    |                         `AudioContextState` is not defined in the current crate
    | |    `ProcessingState` is not defined in the current crate
    | impl doesn't use only types from inside the current crate
    |
    = note: define and implement a trait or new type instead
```

is `error[E0599]: no method named `chars` found for reference `&str` in the current scope` your first error?

**Comment by jdm** [comment]
No, I've been skipping over those orphan rule violations because I understand them. I have not found going top to bottom to be the most useful strategy so far.

**Comment by sagudev** [comment]
> I have not found going top to bottom to be the most useful strategy so far.

Well, it the most sure ways to fix proper error properly. I got burnt few times from fixing wrong errors at wrong time.

**Comment by jdm** [comment]
My branch now a) compiles, b) open pages that don't use any JS. I know what's missing that should get JS working as well.

**Comment by sagudev** [comment]
Plan from https://github.com/servo/servo/pull/33416#issuecomment-2495663559 is already in motion.

**Comment by sagudev** [comment]
We might have problem with rust-analyzer: https://servo.zulipchat.com/#narrow/channel/263398-general/topic/rust-analyzer.20failed.20to.20include.20codes.20in.20script_bindings

**Comment by jdm** [comment]
We're making good progress! #33416 was a +8000 -5000 diff. [My latest work](https://github.com/servo/servo/commit/def3daa05d1a008f8b1cf84b8ef9632c2fc5f514.diff) to see how much is left is only +3500 -3000, and I have a much clearer idea now of which parts can still be moved incrementally, and which parts need to happen all in one PR.

**Comment by mrobinson** [comment]
11 years later and this is closed. Amazing!!
